qutebrowser/utils/utils.py

```diff
@@ -864,6 +864,63 @@ def unused(_arg):
     pass
 
 
+def parse_numsettxt_into_list(txtset, nummax):
```

Comment:
I think we're optimizing the wrong thing here. The thing you want to optimize here is not the total time that it takes to calculate the list, but rather the "response time", which is the time it takes to actually start processing the list. If you do something stupid like `1-100000`, you probably don't want to eagerly create a list of 100000 items, just to notice that you have only 2 downloads in total.

The key here (IMO) is laziness, which can be achieved by using generators instead of lists or smart in-range checks (like Python's `range`). By offloading the total list-creation-time to small steps in between each iteration, we can have a responsive interface with a sub-optimal algorithm, because in many cases we don't even need to create the whole list of all numbers.

This also improves the user interface for those functions. I think a method which returns numbers (such as your first version) is nicer to use than a method for which you have to pass a list of numbers manually. Another nice interface is proper object with a `__contains__` method (like `range`), which would turn

```
ins = utils.which_nums_in_numints(
        range(1, len(self) + 1),
        indexset_p
      )
```

into something like

```
ins = [i for i in range(1, len(self) + 1) if i in indexset_p]
```

At this point you can even iterate over the downloads directly and throw away those that are not in the indexset:

```
for i, download in enumerate(self):
    if i not in indexset_p:
        continue
    ...
```

And if you sacrifice a bit of runtime performance for readability, this `__contains__` method could look like

```
def __contains__(self, item):
    return any(item in r for r in self.ranges)
```

assuming that `self.ranges` is just a list of `range`-objects. That's more readable than a fully-optimized version, while still offering good performance (as long as you don't specify 10000 ranges on the command line).

In the end, it probably depends on the order in which you iterate over the things: do you iterate over all downloads and check if the index is in the indexset, or do you iterate over the indexset and check if there is a corresponding download. The first case is probably nicer, precisely because of cases like `1-1000000`. Other than that, I'd say worrying about runtime complexity is a non-issue here. I doubt that this function will be the bottleneck for anything, and even then, we can still optimize later (after profiling what really takes the most time, as asymptotic runtimes are misleading).

Also, you can sort a list of ranges by using their `.start` attribute, e.g. `list_of_ranges.sort(key=lambda r: r.start)`.